storage/raft_log_queue.go

```
@@ -0,0 +1,141 @@
+	"github.com/cockroachdb/cockroach/gossip"
+	"github.com/cockroachdb/cockroach/roachpb"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/log"
+	"github.com/coreos/etcd/raft"
+)
+
+const (
+	// raftLogQueueMaxSize is the max size of the queue.
+	raftLogQueueMaxSize = 100
+	// RaftLogQueueTimerDuration is the duration between checking the
+	// raft logs.
+	RaftLogQueueTimerDuration = time.Second
+	// RaftLogQueueLogSizeThreshold is the number of raft log entries after
+	// which truncation is required.
+	RaftLogQueueLogSizeThreshold = 1
+)
+
+// raftLogQueue manages a queue of replicas slated to have their raft logs
+// truncated by removing unneeded entries.
+type raftLogQueue struct {
+	baseQueue
+	db *client.DB
+}
+
+// newRaftLogQueue returns a new instance of raftLogQueue.
+func newRaftLogQueue(db *client.DB, gossip *gossip.Gossip) *raftLogQueue {
+	rlq := &raftLogQueue{
+		db: db,
+	}
+	rlq.baseQueue = makeBaseQueue("raftlog", rlq, gossip, raftLogQueueMaxSize)
+	return rlq
+}
+
+func (*raftLogQueue) needsLeaderLease() bool {
+	return false
+}
+
+func (*raftLogQueue) acceptsUnsplitRanges() bool {
+	return true
+}
+
+// getOldestIndex returns the the oldest index still in use for the passed in
+// raft status.
+func getOldestIndex(raftStatus raft.Status) uint64 {
+	// Find the oldest index still in use by the range.
+	oldestIndex := raftStatus.Applied
+	for _, progress := range raftStatus.Progress {
+		if progress.Match < oldestIndex {
+			oldestIndex = progress.Match
+		}
+	}
+
+	return oldestIndex
+}
+
+// shouldQueue determines whether a range should be queued for truncating. This
+// is true only if the replica is the raft leader and if the range's raft log's
+// old entries exceeds the size threshold.
+func (*raftLogQueue) shouldQueue(now roachpb.Timestamp, r *Replica, _ *config.SystemConfig) (shouldQ bool,
+	priority float64) {
+	raftLeader, raftStatus, err := r.isRaftLeader()
+	if !raftLeader {
+		if err != nil {
```

so if `raftLeader` is true, you ignore the error?